### PR TITLE
Publish GTSAM artifacts

### PR DESCRIPTION
**Pull request by varunagrawal**
Updated the Actions workflow to upload the corresponding build directories, so we can use the latest version of GTSAM in CI pipelines for other repos (e.g. GTDynamics).

### CODE REVIEW

File: .github/workflows/build-linux.yml [hunk 1]
```diff
@@ -82,4 +82,9 @@ jobs:
...
         if: runner.os == 'Linux'
         run: |
-          bash .github/scripts/unix.sh -t
\ No newline at end of file
+          bash .github/scripts/unix.sh -t
+      - name: Upload build directory
+        uses: actions/upload-artifact@v2
+        with:
+          name: gtsam-${{ matrix.name }}-${{ matrix.build_type }}
+          path: ${{ github.workspace }}/build/
```
Comment (gchenfc) [inline .github/workflows/build-linux.yml:90]: newline at end of file ?

File: .github/workflows/build-macos.yml [hunk 2]
```diff
@@ -51,3 +51,8 @@ jobs:
         if: runner.os == 'macOS'
         run: |
           bash .github/scripts/unix.sh -t
+      - name: Upload build directory
+        uses: actions/upload-artifact@v2
+        with:
+          name: gtsam-${{ matrix.name }}-${{ matrix.build_type }}
+          path: ${{ github.workspace }}/build/
```
Comment (gchenfc) [inline .github/workflows/build-macos.yml:58]: newline

File: .github/workflows/build-windows.yml [hunk 3]
```diff
@@ -75,4 +75,9 @@ jobs:
...
           cmake --build build --config ${{ matrix.build_type }} --target check.base
           cmake --build build --config ${{ matrix.build_type }} --target check.base_unstable
-          cmake --build build --config ${{ matrix.build_type }} --target check.linear
\ No newline at end of file
+          cmake --build build --config ${{ matrix.build_type }} --target check.linear
+      - name: Upload build directory
+        uses: actions/upload-artifact@v2
+        with:
+          name: gtsam-${{ matrix.name }}-${{ matrix.build_type }}
+          path: ${{ github.workspace }}/build/
```
Comment (gchenfc) [inline .github/workflows/build-windows.yml:83]: newline

**Review by gchenfc — COMMENTED** [review]
Cool!

2 comments:
1. is there a reason not to do this for python and special cases?
2. The artifacts are pretty big since they are uploading the entire build directory, especially the debug zip which is ~10GB.  This has the dual byproducts of extra CI time (~15min for debug to upload the file) and a pretty time to download the artifacts down from github.  Neither of these are deal breakers for me, but just thought I'd point out how large the files are.  Perhaps an argument could be made for a more selective approach?

**Comment by ProfFan** [comment]
@gchenfc Python wheels need very special build environments (check gtsam-manylinux-build) to make them run on every possible Linux distro you have. That is also very expensive (currently needs a custom GitHub runner hosted on GCE).

Also I second Gerry's 2nd point here - I think GitHub has a 10G-ish quota on storage space?

File: .github/workflows/build-linux.yml (hunk 1 above)
Comment (varunagrawal) [reply .github/workflows/build-linux.yml:90]: I checked the file again and it already has a newline.

File: .github/workflows/build-macos.yml (hunk 2 above)
Comment (varunagrawal) [reply .github/workflows/build-macos.yml:58]: Already has a newline :slightly_smiling_face:

File: .github/workflows/build-windows.yml (hunk 3 above)
Comment (varunagrawal) [reply .github/workflows/build-windows.yml:83]: Already has a newline :slightly_smiling_face:

**Comment by varunagrawal** [comment]
I am not sure about the file size limit because Actions is not complaining about it, but you are right, if the upload is adding to the overall time, it doesn't make sense to add all the files. I wonder if zipping up the build directory would make a difference?

File: .github/workflows/build-linux.yml (hunk 1 above)
Comment (gchenfc) [reply .github/workflows/build-linux.yml:90]: sorry my bad!

**Comment by gchenfc** [comment]
[github billing](https://docs.github.com/en/free-pro-team@latest/github/setting-up-and-managing-billing-and-payments-on-github/about-billing-for-github-actions)
Not sure what account borglab is on but this would seem to imply that indeed there are storage limits which will incur additional cost.

Hmm.  Maybe not using debug builds and also only uploading builds on merge with develop?
I think the zipped files are ~1/4 the size of the original directory and take virtually no time to zip

**Comment by varunagrawal** [comment]
> GitHub Actions usage is free for public repositories and self-hosted runners.

No worries here.

**Review by gchenfc — APPROVED** [review]
Ahh, didn't read carefully that the limits/billing only applied to private repos.
Personally I would prefer only enabling this on release builds just to speed up builds, but up to you!

**Comment by varunagrawal** [comment]
I completely agree with you on the Release part. For downstream use, Debug builds for the CI just seem like an unnecessary time sink.

**Comment by varunagrawal** [comment]
Also, "free for public repos" is one of the reasons why I convinced Frank to move from Bitbucket to Github. We have so much tooling available to make life easy now.

**Comment by varunagrawal** [comment]
Okay wow, release builds are less than 500 MB while debug builds are > 4 GB!!! Great recommendation @gchenfc.

**Pull request merged**